server/src/test/java/org/elasticsearch/cluster/metadata/MetadataDeleteIndexServiceTests.java

```diff
@@ -92,6 +110,38 @@ public void testDeleteUnassigned() {
         verify(allocationService).reroute(any(ClusterState.class), any(String.class));
     }
 
+    public void testDeleteBackingIndexForDataStream() {
+        int numBackingIndices = randomIntBetween(2, 5);
+        String dataStreamName = randomAlphaOfLength(6).toLowerCase(Locale.ROOT);
+        ClusterState before = DeleteDataStreamRequestTests.getClusterStateWithDataStreams(
+            List.of(new Tuple<>(dataStreamName, numBackingIndices)), List.of());
+
+        int numIndexToDelete = randomIntBetween(1, numBackingIndices - 1);
+
+        Index indexToDelete = before.metadata().index(DataStream.getBackingIndexName(dataStreamName, numIndexToDelete)).getIndex();
+        ClusterState after = service.deleteIndices(before, Set.of(indexToDelete));
+
+        assertNull(after.metadata().getIndices().get(indexToDelete.getName()));
+        for (int k = 1; k <= numBackingIndices; k++) {
+            if (k != numIndexToDelete) {
+                assertNotNull(after.metadata().getIndices().get(DataStream.getBackingIndexName(dataStreamName, k)));
```

Comment:
I think that instead of looping here just checking that the index that is supposed to be removed is removed and checking that the number of indices is decremented by one is sufficient.